### PR TITLE
Add support for hg (Mercurial)

### PR DESCRIPTION
**Pull request by twaugh**

### CODE REVIEW

File: backvendor/filehash.go [hunk 1]
```diff
@@ -35,10 +40,31 @@ type Hasher interface {
...
+// Hash implements the Hasher interface generically using sha256.
+func (h sha256Hasher) Hash(relativePath, absPath string) (FileHash, error) {
+	f, err := os.Open(absPath)
+	if err != nil {
+		return FileHash(""), errors.Wrapf(err, "hashing %s", absPath)
+	}
+	defer f.Close()
+
+	hash := sha256.New()
+	_, err = io.Copy(hash, f)
```
Comment (mprahl) [inline backvendor/filehash.go:54]: Clever trick

Comment (twaugh) [reply backvendor/filehash.go:54]: Interfaces rock.

File: backvendor/hg.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+	// If any is a semver tag, use that
+	for _, entry := range entries {
+		_, err := semver.NewVersion(entry.Tag)
+		if err == nil {
+			return entry.Tag, nil
+		}
+	}
+
+	// Otherwise just take the first one
+	return entries[0].Tag, nil
```
Comment (mprahl) [inline backvendor/hg.go:139]: What's the reason for returning a tag that doesn't map to the semver format?

File: backvendor/hg_test.go [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+	"golang.org/x/tools/go/vcs"
+)
+
+func TestHgLog(t *testing.T) {
+	defer mockExecCommand()()
+
+	h := hgWorkingTree{
+		anyWorkingTree: anyWorkingTree{
+			Source: &GoSource{},
+			VCS:    vcs.ByCmd("hg"),
```
Comment (mprahl) [inline backvendor/hg_test.go]: Optional: You could use your const of `vcsHg` here

File: backvendor/hg.go (hunk 2 above)
Comment (twaugh) [reply backvendor/hg.go:139]: We know that it starts with a number, or 'v' and then a number, so it's at least plausibly a version tag even if it doesn't strictly conform to the semver spec.

The Describable interface says ReachableTag operates "preferring semver tags", i.e. it's allowed to return a non-semver tag if there's no available semver tag.

Comment (twaugh) [reply backvendor/hg.go:139]: FWIW an example of a version tag that is not strictly a semver tag is v1.0.0beta1 (should be v1.0.0-beta1 to be a semver).

File: backvendor/hg_test.go (hunk 3 above)
Comment (twaugh) [reply backvendor/hg_test.go]: Updated.

File: backvendor/hg.go (hunk 2 above)
Comment (mprahl) [reply backvendor/hg.go:139]: Makes sense. Thanks!

File: backvendor/hg_test.go [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+	defer mockExecCommand()()
+
+	wt := hgWorkingTree{
+		anyWorkingTree: anyWorkingTree{
+			Source: &GoSource{},
+			VCS:    vcs.ByCmd(vcsHg),
+		},
+	}
+
+	mockedStderr = "fatal: not a git repository\n"
```
Comment (mprahl) [inline backvendor/hg_test.go]: This should likely be something mercurial specific.

**Review by mprahl — APPROVED** [review]
:+1:

File: backvendor/hg_test.go (hunk 4 above)
Comment (twaugh) [reply backvendor/hg_test.go]: Fixed.

**Pull request merged**